### PR TITLE
Refactoring syntax analyzer

### PR DESCRIPTION
**Pull request by boki1**
### Description And Task

During the preplanning period for the Two Towers epic, major problems with code base for the syntax analyzer were found. These include: namespace overlapping (or missing concrete), poor OOP design, etc.

A better version is proposed which has the obligation to satisfy the requirements of the design period for the next epic. These include:

- Better building system
    + Problem: Use a unified one for the whole code project
    + Decision: Use CMakeLists

- Better directory structuring
    + Problem: Poor separation of the different code base modules
    + Decision: Use additional directories (with associated building instruction files) for each module

- Lack of tests
    + Problem The code base has 0% code coverage, having tests only for the final result
    + Decision Test it  

- Lack of documentation

    + Problem For the most part the code does not contain additional documentation info
    + Decision Document it  

- Poor OO design
    + Problem Although each component of syntax analyzer is exported in a class on its own, the division is not good enough
    + Decision Extract the components better    

### Definition of Done

A fully functional syntax analyzer (according to the grammar) which successfully performs its tasks and additionally satisfies the said decisions and requirements.

### CODE REVIEW

**Review by vkabadzhova — CHANGES_REQUESTED** [review]
Excellent - very good design and as it can be seen, a lot of work is put in here. Still, some more details should be done, for example we must take into consideration:
[ ] The bug in the examle/gcd
[ ] Refactoring of the m_all_scopes in the sa::parsing_context, as mentioned

File: src/kivi/ast/src/ast/parsing_context.hh [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		//
+
+		/**
+		 * The constructor initializes the "location" variables in both the
+		 * lexer and the parser.
+		 *
+		 * @param code_beginning  the "location" var for the lexer
+		 * @param filename the "location" var for the parser
+		 */
+		parsing_context(const char* code_beginning, std::string* filename)
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.hh]: I do not think `explicit` would be a good addition, since the constructor accepts multiple arguments and I do not think implicit conversion could happen.

File: src/kivi/ast/src/ast/parsing_context.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+namespace syntax_analyzer
+{
+
+	void parsing_context::enter_scope()
+	{
+		LOG_F(INFO, "Emplacing a new scope at the end\n");
+		m_all_scopes.emplace_back();
+	}
+
+	void parsing_context::exit_scope()
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.cc]: I like that. I will immediately apply that.

File: src/kivi/ast/src/ast/parsing_context.hh [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+#include <kivi_parser/location.hh>
+#include <kivi_parser/stack.hh>
+
+#include <kivi_expressions/identifier.hh>
+#include <kivi_stmts/statement.hh>
+
+#include "syntax.hh"
+
+namespace syntax_analyzer
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.hh]: I do not think that is needed since no active usages appear in the current version of the file.

File: src/kivi/ast/src/ast/syntax.hh [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+#include <string>
+#include <utility>
+
+#include <kivi_expressions/base.hh>
+#include <kivi_stmts/statement.hh>
+
+#include "syntactic_structure.hh"
+
+namespace syntax_analyzer
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/syntax.hh]: I do not think that is needed here since no active usages appear in the current version of the file.

File: src/misc/lexemes.re [hunk 5]
```diff
@@ -1,48 +1,54 @@
 #include "kivi_parser.tab.hh"
-#include "../include/parser.hpp"
+#include <ast/parsing_context.hh>
 
-namespace yy {
+// Namespace shortcuts
+namespace sa = syntax_analyzer;
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/misc/lexemes.re:5]: I appreciate your note, but I really do think of `sa` as an _alias_ and the actual name. On modern IDEs (tested on Clion and nvim with `coc`) actually directly connect you with the actual name. This alias is used only in the internal implementation of the parser. In order to provide an actual name - describing and meaningful, I cannot think of anything shorter than `syntax_analyzer` so the alias would not serve any no actual purpose. I actually agree with you on this but I wouldn't like the following if no alias was provided.

```cpp
std::variant<sa::I_statement, sa::function, sa::identifier, sa::I_expression, sa::invalid> m_field;
```  

In conclusion, I do agree that the name `sa` is not detailed enough, but I do still think that the alias serves good enough purpose in the current situation.

**PS:** As a  middle-of-the-road-solution I would suggest that when large declarations (or other) are typed, an alias is used, but when `syntax_analyzer` does not pollute the code it should be preferred.

File: src/kivi/ast/src/ast/parsing_context.cc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	{
+		LOG_F(INFO, "Popping last scope from the back\n");
+		m_all_scopes.pop_back();
+	}
+
+	const identifier&
+	parsing_context::define_identifier(const std::string& name, identifier&& ident)
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.cc]: Thanks. I will take this in account and provided a reasonable fix ASAP.

File: src/kivi/ast/src/ast/parsing_context.cc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	{
+		LOG_S(INFO) << "Defining identifier \"" << name << "\"";
+		auto it = all_scopes_mut().back().emplace(name, std::move(ident));
+
+		if (!it.second)
+		{
+			LOG_S(WARNING) << "Duplicate definition of \"" + ident.name() << "\".";
+			throw kp::syntax_error(yy_location, "Duplicate definition of \"" + ident.name() + "\".");
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.cc]: Is [this](https://github.com/boki1/kivi/pull/15#discussion_r628214387) the so-called below :smile:. If so, as far as I understand the problem (the warnings), the issue is not with us, but the internal implementation of the other libraries (`bison` and `loguro`).

**PS:** Pls, correct me if I am wrong. Looking forward to hearing from you.

File: src/kivi/ast/src/ast/parsing_context.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (vkabadzhova) [reply src/kivi/ast/src/ast/parsing_context.hh]: There should be a way, somehow.. Let's put it in the backlog as a nice-to-have :wink:

File: src/kivi/ast/src/ast/syntax.hh (hunk 4 above)
[the remark this answers is not in the document]
Comment (vkabadzhova) [reply src/kivi/ast/src/ast/syntax.hh]: There are no active usages, but as a library, it is good to have it. Imagine a friend Joe, who wants to use the syntax_analyzer namespace and knows that we have a handy sa alias in the ast lib. Anyway, he knows that in the parsing_context library there is syntax_analyzer namespace, too, so he tries to use sa as it's logically. How would Joe feel?... :disappointed:

So let's add it! :smile:

**Review by vkabadzhova — CHANGES_REQUESTED** [review]
Please, do not commit docs/

File: src/kivi/ast/src/ast/parsing_context.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (vkabadzhova) [reply src/kivi/ast/src/ast/parsing_context.cc]: Yes, that's the exact comment. OK, we'll see if we can do something

File: src/kivi/ast/src/ast/parsing_context.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (vkabadzhova) [reply src/kivi/ast/src/ast/parsing_context.cc]: Thank you!

File: src/kivi/ast/src/ast/parsing_context.cc [hunk 8]
```diff
@@ -23,9 +23,13 @@ namespace syntax_analyzer
 		m_all_scopes.emplace_back();
 	}
 
-	void parsing_context::exit_scope()
+	void parsing_context::exit_scope() noexcept(false)
```
Comment (vkabadzhova) [inline src/kivi/ast/src/ast/parsing_context.cc]: whyyy? why `noexcept(false)`: https://stackoverflow.com/questions/16244313/does-adding-noexceptfalse-benefit-the-code-in-any-way

**Review by vkabadzhova — COMMENTED** [review]
I'll anticipate your comment - look at the Stack Overflow's question I've commented and tell your opinion on if we should use `noexcept(false)`, and then, according to the decision, let's make a KIVI standard :smiley:

File: src/kivi/ast/src/ast/parsing_context.cc (hunk 8 above)
Comment (vkabadzhova) [reply src/kivi/ast/src/ast/parsing_context.cc]: **_Update:_** to be decided

File: src/kivi/ast/src/ast/parsing_context.hh [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+{
+
+	class parsing_context
+	{
+	 private:
+		/// All parsed scopes
+		std::vector<std::map<std::string, identifier>> m_all_scopes;
```
[the remark this answers is not in the document]
Comment (boki1) [reply src/kivi/ast/src/ast/parsing_context.hh]: I fixed it.

**Review by vkabadzhova — APPROVED** [review]
